File: Samples/PISystem/VB.NET/GSharp.PythonRegressionAnalysis/GSharp.PythonRegressionAnalysis/LinearRegression.vb
Imports System.IO
Imports System.Reflection
Imports System.Text.Json
Imports System.Threading
Imports GHost.GSharp.Calculation.AF
Imports GHost.GSharp.Core.AF.Data
Imports GHost.GSharp.Core.Attributes
Imports GHost.GSharp.Core.Enums
Imports GHost.GSharp.Core.Interfaces.Model
Imports GHost.GSharp.PythonNet.Calculation
Imports OSIsoft.AF.Asset
Imports OSIsoft.AF.Time
Imports Python.Runtime

''' <summary>
''' A gSharp execution module.
''' This module will be imported to the server at the time the containing assembly is registered.
''' </summary>
Public Class LinearRegression
    Inherits CalculationModuleAFBase

    Private _PythonHost As PythonHost

    <PointDataDirection(DataDirection.Input)>
    Public Property ProcessFeedrate As AFAttrFloatPoint

    <PointDataDirection(DataDirection.Output)>
    Public Property Intercept As AFAttrFloatPoint

    <PointDataDirection(DataDirection.Output)>
    Public Property Slope As AFAttrFloatPoint


    Public Overrides Sub Initialize()
        _PythonHost = InitializePython()
    End Sub

    Public Overrides Sub Execute(schedule As ISchedule, timestamp As Date, cancelToken As CancellationToken)
        Try
            ' Depending on requirements, data can be retrieved using one of several options:
            ' either the built-in RecordedValues or InterpolatedValues methods of AFAttrxxxPoint.Data,
            ' or using the underlying AFSDK Summaries method to get averaged data

            ' Uncomment the code below to use the RecordedValues method for raw archive data. Change the time range as needed.
            'Dim tsDataset As AFValues = ProcessFeedrate.Data.RecordedValues("*-8h", "*", AFBoundaryType.Inside, Nothing)

            ' Uncomment the code below to use the InterpolatedValues method for evenly spaced interpolated data. Change the time range and sampling interval as needed.
            Dim tsDataset As AFValues = ProcessFeedrate.Data.InterpolatedValues("*-8h", "*", New AFTimeSpan(0, 0, 0, 0, 5), Nothing, Nothing, True)

            ' Uncomment the code below to use averaged data over the time range. Change the time range and calculation interval as needed.
            'Dim tsDataset As New AFValues
            'Dim summariesDataset = ProcessFeedrate.PIPoint.Summaries(New OSIsoft.AF.Time.AFTimeRange("*-8h", "*"),
            '                                                  New OSIsoft.AF.Time.AFTimeSpan(0, 0, 0, 0, 5),
            '                                                  OSIsoft.AF.Data.AFSummaryTypes.Average,
            '                                                  OSIsoft.AF.Data.AFCalculationBasis.TimeWeighted,
            '                                                  OSIsoft.AF.Data.AFTimestampCalculation.Auto)

            'summariesDataset.TryGetValue(OSIsoft.AF.Data.AFSummaryTypes.Average, tsDataset)

            Dim cleanDataSet As List(Of AFValue) = tsDataset.Where(Function(v) v.IsGood).ToList()

            Dim pyDictionary = New JsonDataDictionary(cleanDataSet)
            Dim pyDataset = JsonSerializer.Serialize(pyDictionary)
            Dim code As String = ReadPythonScript("LinearRegression.py")

            Dim pySlope As Object = Nothing
            Dim pyIntercept As Object = Nothing

            Using scope As ModuleScope = _PythonHost.CreateScope()
                Dim pyInputData As PyObject = pyDataset.ToPython()
                scope.Set("data", pyInputData)
                scope.Exec(code)
                scope.TryGet("slope", pySlope)
                scope.TryGet("intercept", pyIntercept)
            End Using

            Dim slopeValue As Single = Convert.ToSingle(pySlope)
            Dim interceptValue As Single = Convert.ToSingle(pyIntercept)

            Intercept.Value = interceptValue
            Slope.Value = slopeValue

        Catch ex As Exception
            Log.Error("An error occurred during Linear Regression calculation: ", ex.Message)
        End Try
    End Sub

    Private Function ReadPythonScript(fileName As String) As String
        Dim assembly As Assembly = assembly.GetExecutingAssembly()
        Dim resourcePath As String = fileName

        resourcePath = assembly.GetManifestResourceNames().Single(Function(str) str.EndsWith(fileName))
        Using stream As Stream = assembly.GetManifestResourceStream(resourcePath)
            Using reader As New StreamReader(stream)
                Return reader.ReadToEnd()
            End Using
        End Using
    End Function

    ' An alternative method for reading the Python script. In this case the script
    ' file would not be an embedded resource, but copied to the output directory on build,
    ' and then deployed with the gSharp assembly and it's other dependencies.
    ' An advantage of this approach is that the Python script could be updated as
    ' needed without recompiling the gSharp calculation assembly.

    'Private Function ReadPythonScript(fileName As String) As String
    '    Dim assembly As Assembly = Assembly.GetExecutingAssembly()
    '    Dim filePath As String = Path.Combine(Path.GetDirectoryName(assembly.Location), "PythonScripts", fileName)

    '    Using stream As Stream = File.OpenRead(filePath)
    '        Using reader As New StreamReader(stream)
    '            Return reader.ReadToEnd()
    '        End Using
    '    End Using
    'End Function

    Public Overrides Sub Dispose()
        ' Place any class-level disposal code in here.
        ' This method is called only once, at the time the context is unscheduled Or disabled.
    End Sub
End Class

Friend Class JsonDataDictionary
    Public Property [Date] As List(Of String)
    Public Property Value As List(Of String)

    Public Sub New(data As List(Of AFValue))
        [Date] = New List(Of String)
        Value = New List(Of String)

        For Each value As AFValue In data
            [Date].Add(value.Timestamp.ToString("yyyy-MM-ddTHH:mm:ss"))
            Me.Value.Add(value.ValueAsDouble().ToString())
        Next
    End Sub
End Class

File: Samples/PISystem/CSharp/GSharp.PythonRegressionAnalysis/GSharp.PythonRegressionAnalysis/LinearRegression.cs
using GHost.Core;
using GHost.Core.Logging.Context;
using GHost.Core.Time;
using GHost.GSharp.Calculation.AF;
using GHost.GSharp.Core.AF.Data;
using GHost.GSharp.Core.AF.Time;
using GHost.GSharp.Core.Attributes;
using GHost.GSharp.Core.Enums;
using GHost.GSharp.Core.Interfaces.Model;
using GHost.GSharp.Core.PI.Enums;
using GHost.GSharp.PythonNet.Calculation;
using OSIsoft.AF.Asset;
using OSIsoft.AF.Data;
using OSIsoft.AF.Time;
using OSIsoft.AF.UnitsOfMeasure;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Threading;

namespace GSharp.PythonRegressionAnalysis;

/// <summary>
/// A gSharp execution module.
/// This module will be imported to the server at the time the containing assembly is registered.
/// </summary>
public class LinearRegression : CalculationModuleAFBase
{
    private PythonHost _PythonHost;

    [PointDataDirection(DataDirection.Input)]
    public AFAttrFloatPoint ProcessFeedrate { get; set; }

    [PointDataDirection(DataDirection.Output)]
    public AFAttrFloatPoint Intercept { get; set; }

    [PointDataDirection(DataDirection.Output)]
    public AFAttrFloatPoint Slope { get; set; }


    public override void Initialize()
    {
        _PythonHost = this.InitializePython();
    }

    public override void Execute(ISchedule schedule, DateTime timestamp, CancellationToken cancelToken)
    {
        try
        {
            /* Depending on requirements, data can be retrieved using one of several options: 
             * either the built-in RecordedValues or SampledValues methods of AFDataPoint,
             * or using the underlying AFSDK Summaries method to get averaged data */

            // Uncomment the code below to use the RecordedValues method for raw archive data. Change the time range as needed.
            //AFValues tsDataset = ProcessFeedrate.Data.RecordedValues("*-8h", "*", AFBoundaryType.Inside, null);

            // Uncomment the code below to use the SampledValues method for evenly spaced interpolated data. Change the time range and sampling interval as needed.
            AFValues tsDataset = ProcessFeedrate.Data.InterpolatedValues("*-8h", "*", new AFTimeSpan(0, 0, 0, 0, 5), null, null, true);

            // Uncomment the code below to use averaged data over the time range. Change the time range and calculation interval as needed.
            //var summariesDataset = ProcessFeedrate.PIPoint.Summaries(new AFTimeRange("*-8h", "*"),
            //    new AFTimeSpan(0, 0, 0, 0, 5), // Represents a 5 minute interval
            //    AFSummaryTypes.Average,
            //    AFCalculationBasis.TimeWeighted,
            //    AFTimestampCalculation.Auto);
            //summariesDataset.TryGetValue(OSIsoft.AF.Data.AFSummaryTypes.Average, out AFValues tsDataset);

            // Clean the data before further processing, i.e., remove any Bad values
            var cleanDataSet = tsDataset.Where(v => v.IsGood).ToList();

            var pyDictionary = new JsonDataDictionary(cleanDataSet);
            string pyDataset = JsonSerializer.Serialize(pyDictionary);
            string code = ReadPythonScript("LinearRegression.py");

            object pySlope;
            object pyIntercept;

            using (ModuleScope scope = _PythonHost.CreateScope())
            {
                scope.Set("data", pyDataset);
                scope.Exec(code);
                scope.TryGet<object>("slope", out pySlope);
                scope.TryGet<object>("intercept", out pyIntercept);
            }

            float.TryParse(pySlope.ToString(), out float slopeValue);
            float.TryParse(pyIntercept.ToString(), out float interceptValue);

            Slope.Value = slopeValue;
            Intercept.Value = interceptValue;
        }
        catch (Exception ex)
        {
            Log.Error($"An error occurred during Linear Regression calculation: {ex.Message}");
        }
    }

    private string ReadPythonScript(string fileName)
    {
        var assembly = System.Reflection.Assembly.GetExecutingAssembly();
        string resourcePath = fileName;
        resourcePath = assembly.GetManifestResourceNames()
            .Single(str => str.EndsWith(fileName));

        using Stream stream = assembly.GetManifestResourceStream(resourcePath);
        using StreamReader reader = new(stream);
        return reader.ReadToEnd();
    }

    //private string ReadPythonScript(string fileName)
    //{
    //    var assembly = System.Reflection.Assembly.GetExecutingAssembly();
    //    string filePath = Path.Combine(Path.GetDirectoryName(assembly.Location), "PythonScripts", fileName);

    //    using Stream stream = File.OpenRead(filePath);
    //    using StreamReader reader = new(stream);
    //    return reader.ReadToEnd();
    //}

    public override void Dispose()
    {
        // Place any class-level disposal code in here.
        // This method is called only once, at the time the context is unscheduled or disabled.
    }
}

internal class JsonDataDictionary
{
    public List<string> Date { get; private set; }

    public List<string> Value { get; private set; }

    public JsonDataDictionary(List<AFValue> data)
    {
        Date = [];
        Value = [];

        foreach (var val in data)
        {
            Date.Add(val.Timestamp.ToString("yyyy-MM-ddTHH:mm:ss"));
            Value.Add(val.ValueAsDouble().ToString());
        }
    }
}
